src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -346,6 +362,58 @@ def __str__(self):
             self._datatype, ", ".join(dims)))
 
 
+class StructureType(DataType):
+    '''
+    Describes a 'structure' or 'derived' datatype that is itself composed
+    of a list of other datatypes. Those datatypes are stored as an
+    OrderedDict of namedtuples.
+
+    '''
+    # Each member of a StructureType is represented by a ComponentType
+    # (named tuple).
+    ComponentType = namedtuple("ComponentType", ["name", "datatype",
+                                                 "visibility"])
+
+    def __init__(self):
+        self._components = OrderedDict()
+
+    def __str__(self):
+        return "StructureType<>"
+
+    @property
+    def components(self):
+        '''
+        :returns: Ordered dictionary of the components of this type.
+        :rtype: :py:class:`collections.OrderedDict`
+        '''
+        return self._components
+
+    def add(self, component):
```

Comment:
Since the namedtuple doesn't have proper type checking, I would prefer this method to be `def add(self, name, datatype, visibility)` which each attribute checked here.

This would also keeps the ComponentType more as an internal implementation detail (although still exposed by the lookup).

Reply:
Yes, that's a good idea. Done.